src/usb/uf2/ghostfat.c

```diff
@@ -288,68 +291,217 @@ void read_block(uint32_t block_no, uint8_t *data) {
+       * TODO May be worth to have some kind crc/application integrity checking
+       *
+       *                         -------------         -------------         -------------
+       *                        |             |       |             |     + |     New     |
+       *                        | Bootloader  |       | Bootloader  |    +  | Bootloader  |
+       * BOOTLOADER_ADDR_START--|-------------|       |-------------|   +   |-------------|
+       *                        |  App Data   |       |  App Data   |  +    |   App Data  |
+       *       USER_FLASH_END --|-------------|       | ----------  | +     |------------ |
+       *                        |             |       |     New     |+      |             |
+       *                        |             | --->  |  Bootloader |       |             |
+       *                        |             |       |   ++++++    |       |             |
+       *                        | Application |       | Application |       | Application |
+       *                        |             |       |             |       |             |
+       *                        |             |       |             |       |             |
+       *      USER_FLASH_START--|-------------|       |-------------|       |-------------|
+       *                        |     MBR     |       |     MBR     |       |     MBR     |
+       *                         -------------         -------------         -------------
+       */
+      PRINTF("addr = 0x%08lX, block = %ld (%ld of %ld)\r\n", bl->targetAddr, bl->blockNo, state->numWritten, bl->numBlocks);
+
+      state->update_bootloader = true;
+      if ( in_uicr_space(bl->targetAddr) )
+      {
+        /* UCIR contains bootloader & MBR address as follow:
+         * - 0x10001014 bootloader address
+         * - 0x10001018 MBR Params: mostly fixed
+         *
+         * Since the bootloader start address is fixed, we only use this for verification
+         */
+        uint32_t uicr_boot_addr;
+        uint32_t uicr_mbr_param;
+
+        memcpy(&uicr_boot_addr, bl->data + 0x14, 4);
+        memcpy(&uicr_mbr_param, bl->data + 0x18, 4);
+
+        // Check MBR params is fixed and prohibited to change and
+        // Bootloader address against its new size
+        if ( (uicr_boot_addr != BOOTLOADER_ADDR_START)  ||
+             (uicr_mbr_param != BOOTLOADER_MBR_PARAMS_PAGE_ADDRESS) )
+        {
+          PRINTF("Incorrect UICR value");
+          PRINT_HEX(uicr_boot_addr);
+          PRINT_HEX(uicr_mbr_param);
+
+          state->aborted = true;
+          return -1;
+        }
 
-        static bool first_write = true;
-        if ( first_write ) {
-          first_write = false;
-          led_state(STATE_WRITING_STARTED);
+        state->has_uicr = true;
+      }
+      else if ( in_bootloader_space(bl->targetAddr) )
+      {
+        // Bootloader CF2 config
+        if ( !state->boot_id_matches && (bl->targetAddr >= ((uint32_t) bootloaderConfig)) )
```

Comment:
this reject the bootloader.uf2 that contain cf2 with incorrect BOOTLOADER_ID (VID/PID).